api/src/services/bctw-service.ts

```diff
@@ -0,0 +1,194 @@
+  manufacturer: string;
+  model: string;
+  attachment_start: string;
+  attachment_end: string;
+  critter_id: string;
+}
+
+export interface IDeploymentUpdate {
+  deployment_id: string;
+  attachment_start: string;
+  attachment_end: string;
+}
+
+export interface IDeploymentRecord {
+  assignment_id: string;
+  collar_id: string;
+  critter_id: string;
+  created_at: string;
+  created_by_user_id: string;
+  updated_at: string;
+  updated_by_user_id: string;
+  valid_from: string;
+  valid_to: string;
+  attachment_start: string;
+  attachment_end: string;
+  deployment_id: number;
+}
+
+export interface IBctwUser {
+  keycloak_guid: string;
+  username: string;
+}
+
+export const BCTW_API_HOST = process.env.BCTW_API_HOST || '';
+export const DEPLOY_DEVICE_ENDPOINT = '/deploy-device';
+export const GET_DEPLOYMENTS_ENDPOINT = '/get-deployments';
+export const UPDATE_DEPLOYMENT_ENDPOINT = '/update-deployment';
+export const GET_COLLAR_VENDORS_ENDPOINT = '/get-collar-vendors';
+export const HEALTH_ENDPOINT = '/health';
+
+export class BctwService {
+  user: IBctwUser;
+  keycloak: KeycloakService;
+
+  constructor(user: IBctwUser) {
+    this.user = user;
+    this.keycloak = new KeycloakService();
+  }
+
+  /**
+   * Return user information as a JSON string.
+   *
+   * @return {*}  {string}
+   * @memberof BctwService
+   */
+  getUserHeader(): string {
+    return JSON.stringify(this.user);
+  }
+
+  /**
+   * Retrieve an authentication token using Keycloak service.
+   *
+   * @return {*}  {Promise<string>}
+   * @memberof BctwService
+   */
+  async getToken(): Promise<string> {
+    const token = await this.keycloak.getKeycloakServiceToken();
+    return token;
+  }
+
+  /**
+   * Throw an error if the request to the BCTW API fails.
+   *
+   * @param {AxiosResponse<any>} response
+   * @param {string} endpoint
+   * @memberof BctwService
+   */
+  handleRequestError(response: AxiosResponse<any>, endpoint: string) {
+    if (!response.data || response.status >= 400) {
+      throw new ApiError(ApiErrorType.UNKNOWN, `API request to ${endpoint} failed with status code ${response.status}`);
+    }
+  }
+
+  /**
+   * Send an authorized get request to the BCTW API.
+   *
+   * @param {string} endpoint
+   * @param {Record<string, string>} [queryParams] - An object containing query parameters as key-value pairs
+   * @return {*}
+   * @memberof BctwService
+   */
+  async makeGetRequest(endpoint: string, queryParams?: Record<string, string>) {
+    let url = `${BCTW_API_HOST}${endpoint}`;
+    const token = await this.getToken();
+
+    // If query parameters are provided, append them to the URL
+    if (queryParams) {
+      const params = new URLSearchParams(queryParams);
+      url += `?${params.toString()}`;
+    }
+
+    const response = await axios.get(url, {
+      headers: {
+        authorization: `Bearer ${token}`,
+        user: this.getUserHeader()
+      }
+    });
+
+    this.handleRequestError(response, url);
```

Comment:
If the request is fulfilled successfully, I don't think it makes a lot of sense to run a `handleRequestError` function that checks if an error occured and only occassionally actually acts on an error. Axios is already able to determine if the request errored. Two ways: if you use the await keyword, you can wrap the `get` call in a `try catch` block; Otherwise, you can just chain `.catch`:

```javascript
try {
  const response = await axios.get(url, headers);
} catch (actualError) {
  this.handleError(url, actualError);
}
```

Chaining method:

```javascript
const response = await axios.get(url, headers).catch((actualError) => this.handleError(url, actualError));
```

Reply:
In general, prefer using await when possible.

There are a few situations when the chaining is needed or makes it simpler, but otherwise await tends to be easier to read and understand.